src/eos-crash-metrics.c

```diff
@@ -137,6 +182,122 @@ normalize_path (char *path)
+{
+  g_autoptr(GBytes) metadata = NULL;
+  gchar *runtime = NULL;
+  g_autoptr(GKeyFile) key_file = NULL;
+  const gchar *metadata_data;
+  gsize metadata_len;
+  g_autoptr(GError) local_error = NULL;
+
+  metadata = flatpak_installed_ref_load_metadata (FLATPAK_INSTALLED_REF (ref), NULL, &local_error);
+
+  if (metadata == NULL)
+    {
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
+    }
+
+  metadata_data = g_bytes_get_data (metadata, &metadata_len);
+  key_file = g_key_file_new ();
+
+  if (!g_key_file_load_from_data (key_file, metadata_data, metadata_len, G_KEY_FILE_NONE, &local_error))
+    {
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
+    }
+
+  if (!(runtime = g_key_file_get_string (key_file, "Application", "runtime", &local_error)))
+    {
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
+    }
+
+  sanitize_name (runtime);
+  return runtime;
+}
+
+static FlatpakInfo *
+get_flatpak_info (const char *path,
+                  GError **error)
+{
+  FlatpakInfo *info;
+  g_autoptr(GPtrArray) xrefs = NULL;
+  g_autoptr(GPtrArray) runtimes = NULL;
+  guint i;
+  g_autofree char *executable_name = NULL;
+  g_autoptr(FlatpakInstallation) installation = flatpak_installation_new_system (NULL, error);
+
+  if (installation == NULL)
+    return NULL;
+
+  executable_name = g_path_get_basename (path);
+
+  /* get apps and runtimes */
+  xrefs = flatpak_installation_list_installed_refs (installation,
+                                                    NULL, error);
+  if (xrefs == NULL)
+    return NULL;
+
+  info = flatpak_info_new ();
+
+  runtimes = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+
+  for (i = 0; i < xrefs->len; i++)
+    {
+      g_autofree gchar *executable_path = NULL;
+      FlatpakInstalledRef *xref = g_ptr_array_index (xrefs, i);
+
+      /* store the runtimes */
+      if (flatpak_ref_get_kind (FLATPAK_REF (xref)) == FLATPAK_REF_KIND_RUNTIME)
+        {
+          g_ptr_array_add (runtimes, g_object_ref(xref));
+          continue;
+        }
+
+      if ((flatpak_ref_get_kind (FLATPAK_REF (xref)) == FLATPAK_REF_KIND_APP) && info->app)
+          continue;
+
+      executable_path = g_build_filename (flatpak_installed_ref_get_deploy_dir (FLATPAK_INSTALLED_REF (xref)),
+                                          "files",
+                                          "bin",
+                                          executable_name,
+                                          NULL);
+      /* found a Flatpak with the same application name */
+      if (g_file_test (executable_path, G_FILE_TEST_EXISTS))
+        info->app = g_object_ref (xref);
+    }
+
+  if (info->app)
+    {
+       g_autofree const char *runtime_name = get_associated_runtime (info->app, error);
+       if (runtime_name == NULL)
+         goto end;
+       for (i = 0; i < runtimes->len; i++)
```

Comment:
Instead of having to iterate over all the runtimes here, since you already have the runtime name, you should be able to use e.g. `flatpak_installation_get_installed_ref()` directly.

Reply:
Indeed! That saves me storing the runtimes as well alltogether and simplifies the loop above as well. I can use "flatpak_installation_list_installed_refs_by_kind" and filter for apps.